lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -1116,6 +1118,191 @@ static bool canonicalizeOrOfConcatsWithCstOperands(OrOp op, size_t concatIdx1,
+    unsigned index;
+    /// Lower bound, if other than -1.
+    std::optional<APInt> lowerBound;
+    /// Upper bound.
+    APInt upperBound;
+
+    /// Returns true if this is an equality check.
+    bool isEqCheck() const {
+      return lowerBound ? (*lowerBound + 2 == upperBound) : upperBound.isOne();
+    }
+  };
+
+  // Identify all the relevant patterns among the inputs,
+  // mark all others to be retained unchanged.
+  auto inputs = op.inputs();
+  llvm::SmallBitVector keptOperands(inputs.size());
+  DenseMap<Value, SmallVector<Interval>> argChecks;
+
+  for (unsigned i = 0, n = inputs.size(); i < n; ++i) {
+    auto input = inputs[i];
+
+    // Find bound checks: and(x > n, x < m)
+    auto andOp = input.getDefiningOp<AndOp>();
+    if (andOp && andOp.inputs().size() == 2) {
+      ICmpOp lhsOp = andOp.inputs()[0].getDefiningOp<ICmpOp>();
+      ICmpOp rhsOp = andOp.inputs()[1].getDefiningOp<ICmpOp>();
+
+      if (lhsOp && rhsOp && lhsOp.lhs() == rhsOp.lhs()) {
+        Value arg = lhsOp.lhs();
+
+        APInt lhsBound, rhsBound;
+        if (matchPattern(lhsOp.rhs(), m_RConstant(lhsBound)) &&
+            matchPattern(rhsOp.rhs(), m_RConstant(rhsBound))) {
+          ICmpPredicate lhsPred = lhsOp.predicate();
+          ICmpPredicate rhsPred = rhsOp.predicate();
+          if (lhsPred == ICmpPredicate::ugt && rhsPred == ICmpPredicate::ult) {
+            argChecks[arg].emplace_back(Interval{i, lhsBound, rhsBound});
+            continue;
+          }
+          if (lhsPred == ICmpPredicate::ult && rhsPred == ICmpPredicate::ugt) {
+            argChecks[arg].emplace_back(Interval{i, rhsBound, lhsBound});
+            continue;
+          }
+        }
+      }
+    }
+
+    if (auto cmpOp = input.getDefiningOp<ICmpOp>()) {
+      APInt v;
+      if (matchPattern(cmpOp.rhs(), m_RConstant(v))) {
+        // Find equality tests: x == n
+        if (cmpOp.predicate() == ICmpPredicate::eq) {
+          if (v.isZero())
+            argChecks[cmpOp.lhs()].emplace_back(
+                Interval{i, std::nullopt, v + 1});
+          else
+            argChecks[cmpOp.lhs()].emplace_back(Interval{i, v - 1, v + 1});
+          continue;
+        }
+        // Find upper bound tests: x < n
+        if (cmpOp.predicate() == ICmpPredicate::ult) {
+          argChecks[cmpOp.lhs()].emplace_back(Interval{i, std::nullopt, v});
+          continue;
+        }
+      }
+    }
+
+    keptOperands[i] = true;
+  }
+
+  // For each value, try to compress the associated checks.
+  llvm::SmallVector<Value> newOperands;
+  for (auto &[arg, checks] : argChecks) {
+    // Order the checks by their lower bounds.
+    std::stable_sort(checks.begin(), checks.end(), [&](auto &lhs, auto &rhs) {
+      if (auto lowerLHS = lhs.lowerBound)
+        if (auto lowerRHS = rhs.lowerBound)
+          return lowerLHS->ult(*lowerRHS);
+      return false;
+    });
+
+    // Find sequences of overlapping checks and compress them.
```

Comment:
This is getting kind of long, might be worth splitting interval merging into a function.

Reply:
The inner loop must access not only the list of checks, but also the array of ups to keep unchanged while producing results - I'd keep it in a single function otherwise the separate calee would have a fairly hard-to-read type.